aws/resource_aws_cloudwatch_log_destination.go

```diff
@@ -64,28 +62,34 @@ func resourceAwsCloudWatchLogDestinationPut(d *schema.ResourceData, meta interfa
 		TargetArn:       aws.String(target_arn),
 	}
 
-	return resource.Retry(3*time.Minute, func() *resource.RetryError {
-		resp, err := conn.PutDestination(params)
+	var resp *cloudwatchlogs.PutDestinationOutput
+	var err error
+	err = resource.Retry(3*time.Minute, func() *resource.RetryError {
+		resp, err = conn.PutDestination(params)
 
-		if err == nil {
-			d.SetId(name)
-			d.Set("arn", *resp.Destination.Arn)
-		}
-
-		awsErr, ok := err.(awserr.Error)
-		if !ok {
+		if isAWSErr(err, cloudwatchlogs.ErrCodeInvalidParameterException, "Could not deliver test message to specified") {
 			return resource.RetryableError(err)
 		}
-
-		if awsErr.Code() == "InvalidParameterException" {
-			if strings.Contains(awsErr.Message(), "Could not deliver test message to specified") {
-				return resource.RetryableError(err)
-			}
+		if isAWSErr(err, cloudwatchlogs.ErrCodeInvalidParameterException, "") {
 			return resource.NonRetryableError(err)
 		}
-
-		return resource.NonRetryableError(err)
+		if isAWSErr(err, "", "") {
+			return resource.NonRetryableError(err)
+		}
+		if err != nil {
+			return resource.RetryableError(err)
+		}
+		return nil
 	})
+	if isResourceTimeoutError(err) {
+		resp, err = conn.PutDestination(params)
+	}
+	if err != nil {
+		return fmt.Errorf("Error putting cloudwatch log destination: %s", err)
+	}
+	d.SetId(name)
+	d.Set("arn", *resp.Destination.Arn)
```

Comment:
Dereferencing types via `*` is extraneous when calling `d.Set()` as it will automatically handle pointer safety. 👍 

```suggestion
	d.Set("arn", resp.Destination.Arn)
```